src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -223,64 +231,193 @@ private UUID getParticipantEntitySetId( UUID studyId ) {
+            DataIntegrationApi dataIntegrationApi ) {
 
-        UUID participantEntityKeyId = getParticipantEntityKeyId( participantId, studyId );
-        if (participantEntityKeyId == null) {
-            logger.error("Unable to retrieve participantEntityKeyId, studyId = {}, participantId = {}", studyId, participantId);
-            return 0;
-        }
+        Map<UUID, Set<Object>> data = new HashMap<>( recordedByEntity );
+        data.put( appPackageNamePropertyTypeId, Sets.newHashSet( appPackageName ) );
 
-        ParticipationStatus status = getParticipationStatus( studyId, participantId );
-        if ( ParticipationStatus.NOT_ENROLLED.equals( status ) ) {
-            logger.warn( "participantId = {} is not enrolled, ignoring data upload", participantId );
-            return 0;
+        return reserveEntityKeyId(
+                recordedByEntitySetId,
+                ImmutableList.of( dateLoggedPropertyTypeId, stringIdPropertyTypeId, appPackageNamePropertyTypeId ),
+                data,
+                dataIntegrationApi
+        );
+    }
+
+    private UUID reserveUserAppEntityKeyId(
+            Map<UUID, Set<Object>> entityData,
+            DataIntegrationApi dataIntegrationApi ) {
+
+        return reserveEntityKeyId(
+                userAppsEntitySetId,
+                ImmutableList.of( appPackageNamePropertyTypeId ),
+                entityData,
+                dataIntegrationApi
+        );
+    }
+
+    private void createUserAppsEntitiesAndAssociations(
+            DataApi dataApi,
+            DataIntegrationApi dataIntegrationApi,
+            List<SetMultimap<UUID, Object>> data,
+            UUID deviceEntityKeyId,
+            UUID participantEntitySetId,
+            UUID participantEntityKeyId,
+            String participantId,
+            String deviceId ) {
+
+        /*
+         * Most of the data pushed by devices does not correspond to apps that were visible in the UI.
+         * Ideally we should only record UsageStats for apps that the user actually used
+         * Method getTotalTimeVisible() available in API level 29 would be the ideal filter since it gives the
+         * total time a package's activity was visible in the UI. However, our android app currently supports minimum API
+         * level 21. so the next best option would be getTotalTimeInForeground() which gives the total time spent in the
+         * foreground, measured in milliseconds
+         *
+         * https://developer.android.com/reference/android/app/usage/UsageStats
+         */
+        List<SetMultimap<UUID, Object>> userAppsData = data
+                .stream()
+                .filter( entry -> entry.containsKey( durationPropertyTypeId ) &&
+                        entry.containsKey( startDateTimePropertyId ) &&
+                        entry.containsKey( appNamePropertyTypeId ) &&
+                        Long.parseLong( entry.get( durationPropertyTypeId ).iterator().next().toString() ) > 0 )
```

Comment:
Why is it important to check `durationPropertyTypeId`? From what I can see, `durationPropertyTypeId` is not used anywhere else except here as part of the conditional expression. Does it need to exist? Is it being used for anything?

Also, I think this is a potential crash point. `Long.parseLong` will throw `NumberFormatException` if the string does not contain a parsable long.

Reply:
The value of `durationPropertyTypeId ` property helps to filter out most of the data that does not correspond to apps that the participant didn't actually use. In the android app, the durationProperty value is set to total time an activity spends in the foreground, measured in milliseconds. This filter is not ideal, but with our current `minSdkVersion` of 21, it's probably the next best option. At API 29, the android sdk has a more fine-grained method of determining which usage stat corresponds to an activity that was visible in the UI.